concur-core/src/Concur/Core/Types.hs

```diff
@@ -1,102 +1,109 @@
 {-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE ExistentialQuantification  #-}
 {-# LANGUAGE FlexibleInstances          #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
 {-# LANGUAGE TupleSections              #-}
 module Concur.Core.Types
   ( Widget(..)
   , continue
-  , widget
   , display
   , mapView
   , wrapView
-  , Suspend(..)
   , SuspendF(..)
-  , Effect
-  , effect
   , awaitViewAction
   , MultiAlternative(..)
   , loadWithIO
   , remoteWidget
   , unsafeBlockingIO
+  , MonadUnsafeBlockingIO(..)
+  , MonadSafeBlockingIO(..)
   ) where
 
-import           Concur.Core.Notify       (Notify, await, newNotify, newNotifyIO, notify)
 import           Control.Applicative      (Alternative, empty, (<|>))
-import           Control.Concurrent       (forkIO)
-import           Control.Concurrent.STM   (STM, atomically, retry)
-import           Control.Monad            (MonadPlus (..))
+import           Control.Concurrent       (MVar, ThreadId, forkIO, killThread, threadDelay, newEmptyMVar, putMVar, takeMVar)
+import           Control.Monad            (MonadPlus (..), forM, forM_)
 import           Control.Monad.Free       (Free (..), hoistFree, liftF)
 import           Control.Monad.IO.Class   (MonadIO, liftIO)
-import           Control.MonadSTM         (MonadSTM (..))
 import           Control.MultiAlternative (MultiAlternative, orr, never)
 
-newtype Widget v a = Widget { suspend :: Free (Suspend v) a }
+import           Concur.Core.Notify
+
+data SuspendF v next
+  = StepView v next
+  | forall r. StepBlock (IO r) (r -> next)
+  | forall r. StepIO    (IO r) (r -> next)
+
+deriving instance Functor (SuspendF v)
+
+newtype Widget v a = Widget { step :: Free (SuspendF v) a }
   deriving (Functor, Applicative, Monad)
 
-data SuspendF v a = SuspendF { view :: v, cont :: Effect a }
-  deriving Functor
+view :: v -> Widget v ()
+view v = Widget $ liftF $ StepView v ()
 
-newtype Suspend v a = Suspend { unSuspend :: IO (SuspendF v a) }
-  deriving Functor
+effect :: IO a -> Widget v a
+effect a = Widget $ liftF $ StepBlock a id
 
-type Effect a = STM (Maybe a)
+io :: IO a ->  Widget v a
+io a = Widget $ liftF $ StepIO a id
 
-continue :: Suspend v a -> Widget v a
+continue :: SuspendF v a -> Widget v a
 continue = Widget . liftF
 
-widget :: v -> Effect a -> Widget v a
-widget v r = continue $ Suspend $ return $ SuspendF v r
-
 display :: v -> Widget v a
-display v = widget v retry
+display v = do
+  view v
+  effect $ do
+    threadDelay maxBound
+    undefined
```

Comment:
I've also been thinking about introducing an additional `Forever` constructor so that we don't spawn a new thread at all in `orr`. Currently every static `text` element has its own thread.

Also, I think it should also be possible to not spawn threads for single children elements (like `button [ text "OK" ]`).